garden-backend-service/src/auth/auth_state.py

```diff
@@ -0,0 +1,71 @@
+import typing as t
+import uuid
+
+import globus_sdk
+from fastapi import HTTPException
+from src.auth.globus_auth import introspect_token
+from src.config import settings
+
+
+class AuthenticationState:
+    """
+    This is a dedicated object for handling authentication.
+
+    It takes in a Globus Auth token and resolve it to a user and various data about that
+    user. It is the "auth_state" object for the application within the context of a
+    request, showing "who" is calling the application (e.g. identity_id) and some
+    information about "how" the call is being made (e.g. scopes).
+
+    For the most part, this should not handle authorization checks, to maintain
+    separation of concerns.
+
+    NB: This class was lifted almost verbatim from funcx web service's auth
+    code. It seemed sane enough, plus doing so will hopefully streamline
+    the Developer Experience of plagiarizing other pieces of their code.
+    """
+
+    def __init__(
+        self, token: t.Optional[str], *, assert_default_scope: bool = True
+    ) -> None:
+        self.garden_default_scope: str = settings.GARDEN_DEFAULT_SCOPE
+        self.token = token
+
+        self.introspect_data: t.Optional[globus_sdk.GlobusHTTPResponse] = None
+        self.identity_id: t.Optional[uuid.UUID] = None
+        self.username: t.Optional[str] = None
+        self.scopes: t.Set[str] = set()
+
+        if token:
+            self._handle_token()
```

Comment:
Are there cases in our app where this is constructed without a token? Just on the lookout for any potential states we could cut out. (Then we wouldn't need `.is_authenticated` for example.)